crates/sui-indexer/src/store/pg_indexer_store_v2.rs

```diff
@@ -512,6 +528,334 @@ impl PgIndexerStoreV2 {
+                    query = query.filter(objects::dsl::object_id.gt(object_cursor.to_vec()));
+                }
+                query.load::<StoredObject>(conn)
+            })
+            .context("Failed to read stored objects from PostgresDB")?;
+            objects
+                .into_iter()
+                .map(|object| object.try_into_object_read(&this.module_cache))
+                .collect::<Result<Vec<_>, _>>()
+        })
+        .await
+    }
+
+    pub async fn get_dynamic_fields(
+        &self,
+        parent_object_id: ObjectID,
+        cursor: Option<ObjectID>,
+        limit: usize,
+    ) -> Result<Vec<DynamicFieldInfo>, IndexerError> {
+        self.execute_in_blocking_worker(move |this| {
+            let objects: Vec<StoredObject> = read_only_blocking!(&this.blocking_cp, |conn| {
+                let mut query = objects::dsl::objects
+                    .filter(objects::dsl::owner_type.eq(OwnerType::Object as i16))
+                    .filter(objects::dsl::owner_id.eq(parent_object_id.to_vec()))
+                    .order(objects::dsl::object_id.asc())
+                    .limit(limit as i64)
+                    .into_boxed();
+                if let Some(object_cursor) = cursor {
+                    query = query.filter(objects::dsl::object_id.ge(object_cursor.to_vec()));
+                }
+                query.load::<StoredObject>(conn)
+            })
+            .context("Failed to read stored objects from PostgresDB")?;
+
+            if any(objects.iter(), |o| o.df_object_id.is_none()) {
+                return Err(IndexerError::PersistentStorageDataCorruptionError(format!(
+                    "Dynamic field has empty df_object_id column for parent object {}",
+                    parent_object_id
+                )));
+            }
+            // for Dynamic field objects, df_object_id != object_id, we need another look up
+            // to get the version and digests.
+            // TODO: simply store df_object_version and df_object_digest as well?
+            let dfo_ids = objects
+                .iter()
+                .filter_map(|o| {
+                    // Unwrap safe: checked nullity above
+                    if o.df_object_id.as_ref().unwrap() == &o.object_id {
+                        None
+                    } else {
+                        Some(o.df_object_id.clone().unwrap())
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            let object_refs = this.get_object_refs(dfo_ids)?;
+            let mut dynamic_fields = objects
+                .into_iter()
+                .map(StoredObject::try_into_expectant_dynamic_field_info)
+                .collect::<Result<Vec<_>, _>>()?;
+
+            for mut df in dynamic_fields.iter_mut() {
+                if let Some(obj_ref) = object_refs.get(&df.object_id) {
+                    df.version = obj_ref.1;
+                    df.digest = obj_ref.2;
+                }
+            }
```

Comment:
I really don't like this part. should have stored df_version and df_digest as well.